akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -1428,3 +1427,63 @@ private[akka] trait CallbackWrapper[T] extends AsyncCallback[T] {
     try body finally lock.unlock()
   }
 }
+
+abstract class AckedCallback[+T](val ack: Option[Promise[_]])
+
+abstract class ConcurrentGraphStageLogic[T] private[stream] (override val inCount: Int, override val outCount: Int) extends GraphStageLogic(inCount, outCount) {
+  def this(shape: Shape) = this(shape.inlets.size, shape.outlets.size)
+
+  val concurrentSet = ConcurrentHashMap.newKeySet[Promise[_]]()
+
+  private trait CallbackState
+  private case class NotInitialized(list: List[AckedCallback[T]]) extends CallbackState
+  private case class Initialized(f: (AckedCallback[T]) ⇒ Unit) extends CallbackState
+  private case class Stopped(f: (AckedCallback[T]) ⇒ Unit) extends CallbackState
+
+  /*
+   * To preserve message order when switching between not initialized / initialized states
+   * lock is used. Case is similar to RepointableActorRef
+   */
+  private[this] final val lock = new ReentrantLock
+
+  private[this] val callbackState = new AtomicReference[CallbackState](NotInitialized(Nil))
+
+  def stopWithStoppedCallbackLogic(f: (AckedCallback[T]) ⇒ Unit): Unit = locked {
+    callbackState.set(Stopped(f))
+  }
+
+  def startWithRunCallbackLogic(f: AckedCallback[T] ⇒ Unit): Unit = locked {
+    val list = (callbackState.getAndSet(Initialized(f)): @unchecked) match {
+      case NotInitialized(l) ⇒ l
+    }
+    list.reverse.foreach(ack ⇒ {
+      ack.ack.map(concurrentSet.add)
+      f(ack)
+    })
+  }
+
+  def asyncInvoke(ac: AckedCallback[T]): Unit = locked {
+    import akka.dispatch.ExecutionContexts.{ sameThreadExecutionContext ⇒ ec }
+
+    ac.ack.map(ack ⇒ {
+      ack.future.andThen { case _ ⇒ concurrentSet.remove(ack) }(ec)
+      concurrentSet.add(ack)
```

Comment:
place this above previous line, so that things are not optimized and remove happens before add